packages/xhr-http-handler/src/xhr-http-handler.ts

```diff
@@ -0,0 +1,222 @@
+
+  /**
+   * Will be called with the xhr instance before calling xhr.send.
+   * Can be used to attach additional event listeners.
+   */
+  xhrBeforeSend?: (xhr: XMLHttpRequest) => void;
+}
+
+/**
+ * An implementation of HttpHandler that uses XMLHttpRequest, which is
+ * traditionally associated with browsers.
+ */
+export class XhrHttpHandler extends EventEmitter implements HttpHandler {
+  /**
+   * Events emitted as an EventEmitter.
+   */
+  public static EVENTS = {
+    /**
+     * Emitted for xhr on progress.
+     * Payload is the native ProgressEvent.
+     */
+    PROGRESS: "progress",
+    /**
+     * Emitted for xhr.upload on progress.
+     * Payload is the native ProgressEvent.
+     */
+    UPLOAD_PROGRESS: "upload.progress",
+  };
+
+  private config?: XhrHttpHandlerOptions;
+  private readonly configProvider?: Provider<XhrHttpHandlerOptions>;
+
+  public constructor(options?: XhrHttpHandlerOptions | Provider<XhrHttpHandlerOptions | undefined>) {
+    super();
+    if (typeof options === "function") {
+      this.configProvider = async () => (await options()) || {};
+    } else {
+      this.config = options ?? {};
+    }
+  }
+
+  public destroy(): void {
+    // Do nothing. Connection pooling is handled by the browser.
+  }
+
+  public async handle(
+    request: HttpRequest,
+    { abortSignal }: HttpHandlerOptions = {}
+  ): Promise<{ response: HttpResponse }> {
+    if (!this.config && this.configProvider) {
+      this.config = await this.configProvider();
+    }
+    const requestTimeoutInMs = Number(this.config!.requestTimeout) | 0;
+
+    // if the request was already aborted, prevent doing extra work
+    if (abortSignal?.aborted) {
+      const abortError = new Error("Request aborted");
+      abortError.name = "AbortError";
+      throw abortError;
+    }
+
+    let path: string = request.path;
+    if (request.query) {
+      const queryString = buildQueryString(request.query);
+      if (queryString) {
+        path += `?${queryString}`;
+      }
+    }
+
+    const { port, method } = request;
+    const url = `${request.protocol}//${request.hostname}${port ? `:${port}` : ""}${path}`;
+    const body = method === "GET" || method === "HEAD" ? undefined : request.body;
+
+    const xhr: XMLHttpRequest =
+      typeof this.config!.xhrFactory === "function" ? await this.config!.xhrFactory() : new XMLHttpRequest();
+
+    const raceOfPromises: Promise<{ response: HttpResponse }>[] = [
+      new Promise((resolve, reject) => {
+        xhr.upload.addEventListener("progress", (event: ProgressEvent) => {
+          this.emit(XhrHttpHandler.EVENTS.UPLOAD_PROGRESS, event, request);
+        });
+        xhr.addEventListener("progress", (event: ProgressEvent) => {
+          this.emit(XhrHttpHandler.EVENTS.PROGRESS, event, request);
+        });
+        xhr.addEventListener('error', (error) => {
+          reject(error);
+        })
+        xhr.addEventListener("readystatechange", () => {
```

Comment:
I don't think the connection timeout(timeout when connection idle too long) is supported here but I think we should. In general, the v2 XHR handler should provide some reference here: https://github.com/aws/aws-sdk-js/blob/master/lib/http/xhr.js